src/System.Console/src/System/IO/StdInStreamReader.cs

```diff
@@ -0,0 +1,350 @@
+using System.Runtime.InteropServices;
+
+namespace System.IO
+{
+    /* This class is used by for reading from the stdin.
+     * It is designed to read stdin in raw mode for interpreting
+     * key press events and maintain its own buffer for the same.
+     * which is then used for all the Read operations
+     */
+    internal class StdInStreamReader : StreamReader
+    {
+        char[] unprocessedBufferToBeRead; // Buffer that might have already been read from stdin but not yet processed.
+        const int bytesToBeRead = 255; // No. of bytes to be read from the stream at a time.
+        int startIndex; // First unprocessed index in the buffer;
+        int endIndex; // Last unprocessed index in the buffer;
+
+        internal StdInStreamReader(Stream stream, Encoding encoding, bool detectEncodingFromByteOrderMarks, int bufferSize, bool leaveOpen) : base(stream, encoding, detectEncodingFromByteOrderMarks, bufferSize, leaveOpen)
+        {
+            unprocessedBufferToBeRead = new char[encoding.GetMaxCharCount(bufferSize)];
+            startIndex = endIndex = -1; // There is no unprocessed index yet.
+        }
+
+        /// <summary> Checks whether the unprocessed buffer is empty. </summary>
+        bool IsExtraBufferEmpty()
+        {
+            if (startIndex == -1 && endIndex == -1) return true;
+
+            if (startIndex > endIndex) // Everything has been processed;
+            {
+                startIndex = -1;
+                endIndex = -1;
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary> Reads the buffer in the raw mode. </summary>
+        unsafe void ReadExtraBuffer()
+        {
+            Debug.Assert(IsExtraBufferEmpty());
+
+            byte[] buffer = new byte[bytesToBeRead];
+            fixed (byte* bufPtr = buffer)
+            {
+                int result;
+                Interop.CheckIo(result = Interop.Sys.ReadStdinUnbuffered((byte*)bufPtr, bytesToBeRead));
+                Debug.Assert(result <= bytesToBeRead);
+
+                // Now we need to convert the byte buffer to char buffer.
+                int charCount = CurrentEncoding.GetMaxCharCount(result);
+                Debug.Assert(charCount < unprocessedBufferToBeRead.Length);
+
+                CurrentEncoding.GetChars(buffer, 0, buffer.Length, unprocessedBufferToBeRead, ++startIndex);
+                endIndex += charCount;
+
+            }
+        }
+
+        public override string ReadLine()
+        {
+            StringBuilder sb = new StringBuilder();
+
+            //Check if there is anything left in the unprocessedBufferToBeRead.
+            while (!IsExtraBufferEmpty())
+            {
+                ConsoleKeyInfo keyInfo = ReadKeyInternal();
+
+                if (keyInfo.Key == ConsoleKey.Enter)
+                    return sb.ToString();
+
+                else if (keyInfo.Key == ConsoleKey.Backspace && sb.Length > 0)
+                    sb.Length--;
```

Comment:
Seems like this would be slightly more efficient as:
```C#
else if (keyInfo.Key == ConsoleKey.Backspace)
{
    int len = sb.Length;
    if (len > 0)
        sb.Length = len-1;
}
```
That way in the case of a backspace we're not checking the ```else if``` below in the case where the StringBuilder was empty.